analysis/builders.go

```diff
@@ -0,0 +1,85 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analysis
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/awslabs/ar-go-tools/analysis/config"
+	"github.com/awslabs/ar-go-tools/analysis/dataflow"
+	"github.com/awslabs/ar-go-tools/analysis/loadprogram"
+	"github.com/awslabs/ar-go-tools/analysis/ptr"
+	"github.com/awslabs/ar-go-tools/internal/formatutil"
+)
+
+// BuildWholeProgramTarget loads the target specified by the list of files provided. Return an analyzer state that has
+// been initialized with the program if successful. The Target of that state will be set to the provided name.
+func BuildWholeProgramTarget(
```

Comment:
Having tried to not have the `BuildXTarget` initially, I found myself just copy-pasting code everywhere that basically implements the `BuildXTarget`. 
The difference between the `NewState` functions for every state and the `BuildXTarget` is:
- with `BuildXTarget` you give files, load options and config and you get `X`. `BuildXTarget` does all the work.
- with `NewState` you have to build the program yourself, or the previous state yourself. `NewState` only builds the additional information on top of its arguments, so you have to chain `loadprogram.Do` and `NewState` for every state you want. This gives more control, but in most places (tests, executables) you just want `BuildXTarget`. `NewState` will be more useful when/if we don't specify a tool and `argot` just detects what analyses it needs to run and what states it needs to build.

Reply:
That's fair. How about renaming each `BuildXTarget` function to `<package>.NewDefaultState`?

Reply:
There is still a difference that the `NewDefaultState` accepts an already built program as input (because it's really here only for tests) whereas the `BuildXTarget` takes in patterns as input, and then builds the program.

Reply:
Ended up removing the `BuildXTarget` functions an instead introducing a result monad to chain the state building operations. Still have some cleanup to do, but this simplifies the interface.